Zend/zend_API.c

```diff
@@ -2003,9 +2003,17 @@ ZEND_API int zend_register_functions(zend_class_entry *scope, const zend_functio
 			}
 			if (info->type_hint) {
 				if (info->class_name) {
+					uint32_t fetch_type;
+					zend_string *class_name;
+					ALLOCA_FLAG(use_heap);
+
 					ZEND_ASSERT(info->type_hint == IS_OBJECT);
-					if (!strcasecmp(info->class_name, "self") && !scope) {
-						zend_error(E_CORE_ERROR, "Cannot declare a return type of self outside of a class scope");
+					STR_ALLOCA_INIT(class_name, info->class_name, strlen(info->class_name), use_heap);
+					fetch_type = zend_get_class_fetch_type(class_name);
+					STR_ALLOCA_FREE(class_name, use_heap);
```

Comment:
hmm, indeed, it's kinda heavy, I just want to avoid duplicate class type checking logic. it was implemented in `zend_get_class_fetch_type`.